### PR TITLE
Issue #2718835 by skyredwang, mglaman: Make ProductAttribute getValue…

### PR DESCRIPTION
**Pull request by mglaman**
… sorted by weights

### CODE REVIEW

**Comment by mglaman** [comment]
@bojanz ready for your eyes. Added kernel test to ensure storage is working as expected.

File: modules/product/src/ProductAttributeValueStorage.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace Drupal\commerce_product;
+
+use \Drupal\commerce\CommerceContentEntityStorage;
+
+/**
+ * Defines storage handler for product attribute value entities.
```
Comment (bojanz) [inline modules/product/src/ProductAttributeValueStorage.php]: "Defines the product attribute value storage."

File: modules/product/src/ProductAttributeValueStorage.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class ProductAttributeValueStorage extends CommerceContentEntityStorage implements ProductAttributeValueStorageInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function loadByPropertiesSortByWeight(array $values = []) {
```
Comment (bojanz) [inline modules/product/src/ProductAttributeValueStorage.php]: Maybe loadByAttributeSortByWeight? Allows us to skip using buildPropertyQuery() for cleaner code

We could probably just call it loadByAttribute().

File: modules/product/src/ProductAttributeValueStorageInterface.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace Drupal\commerce_product;
+
+use Drupal\Core\Entity\ContentEntityStorageInterface;
+
+/**
+ * Defines an interface for product attribute value entity storage classes.
```
Comment (bojanz) [inline modules/product/src/ProductAttributeValueStorageInterface.php]: Defines the interface for product attribute value storage.

File: modules/product/src/ProductAttributeValueStorageInterface.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ */
+interface ProductAttributeValueStorageInterface extends ContentEntityStorageInterface {
+
+  /**
+   * Load entities by their property values and sort by their weights.
```
Comment (bojanz) [inline modules/product/src/ProductAttributeValueStorageInterface.php]: Loads product attribute values for the given product attribute.

Then in @return indicate that the results are sorted by weight.

File: modules/product/tests/src/Kernel/ProductAttributeValueStorageTest.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    $value = array_shift($attribute_values);
+    $this->assertEquals('Cyan', $value->getName());
+    $value = array_shift($attribute_values);
+    $this->assertEquals('Magenta', $value->getName());
+    $value = array_shift($attribute_values);
+    $this->assertEquals('Yellow', $value->getName());
+    $value = array_shift($attribute_values);
+    $this->assertEquals('Black', $value->getName());
+
```
Comment (bojanz) [inline modules/product/tests/src/Kernel/ProductAttributeValueStorageTest.php]: Extra newline.

**Comment by mglaman** [comment]
@bojanz updated

**Comment by bojanz** [comment]
Code looks good, we just need the tests passing.

**Comment by mglaman** [comment]
There's something where the order on save and then load is different when deleting items.

File: modules/product/tests/src/FunctionalJavascript/ProductAttributeJavascriptTest.php [hunk 6]
```diff
@@ -73,14 +73,12 @@ public function testProductAttributeValues() {
...
-    $attribute_values = [];
-    /** @var \Drupal\commerce_product\Entity\ProductAttributeValueInterface $value */
-    foreach ($attribute->getValues() as $value) {
-      $attribute_values[] = $value->label();
-    }
-    $this->assertTrue($attribute_values[0] == 'Cyan');
-    $this->assertTrue($attribute_values[1] == 'Magenta');
-    $this->assertTrue($attribute_values[2] == 'Cornflower Blue');
+    /** @var \Drupal\commerce_product\Entity\ProductAttributeValueInterface[] $attribute_values */
+    $attribute_values = $attribute->getValues();
```
Comment (mglaman) [inline modules/product/tests/src/FunctionalJavascript/ProductAttributeJavascriptTest.php]: I'm stupid. This is keyed by entity ID. Above I properly foreach the results to re-key by weight value.

**Comment by mglaman** [comment]
Holy cow that was evil. The problem was how weight data was being handled in the form (basically holding onto user input when it was invalid.)

**Comment by bojanz** [comment]
Failed on CS. Continued in #465.

**Pull request closed without merging**